frontend/src/scenes/funnels/funnelUtils.ts

```diff
@@ -86,3 +94,67 @@ export function cleanBinResult(binsResult: FunnelsTimeConversionBins): FunnelsTi
         average_conversion_time: binsResult.average_conversion_time ?? 0,
     }
 }
+
+export function aggregateBreakdownResult(
+    breakdownList: FunnelStep[][],
+    breakdownProperty?: string | number | number[]
+): FunnelStepWithNestedBreakdown[] {
+    if (breakdownList.length) {
+        return breakdownList[0].map((step, i) => ({
+            ...step,
+            count: breakdownList.reduce((total, breakdownSteps) => total + breakdownSteps[i].count, 0),
+            breakdown: breakdownProperty,
+            nested_breakdown: breakdownList.reduce(
+                (allEntries, breakdownSteps) => [...allEntries, breakdownSteps[i]],
+                []
+            ),
+            average_conversion_time: null,
+            people: [],
+        }))
+    }
+    return []
+}
+
+export function isBreakdownFunnelResults(results: FunnelStep[] | FunnelStep[][]): results is FunnelStep[][] {
+    return Array.isArray(results) && (results.length === 0 || Array.isArray(results[0]))
+}
+
+// breakdown parameter could be a string (property breakdown) or object/number (list of cohort ids)
+export function isValidBreakdownParameter(breakdown: FunnelRequestParams['breakdown']): boolean {
+    return ['string', 'null', 'undefined', 'number'].includes(typeof breakdown) || Array.isArray(breakdown)
+}
+
+export function wait(ms = 1000): Promise<any> {
+    return new Promise((resolve) => {
+        setTimeout(resolve, ms)
+    })
+}
+
+export const SECONDS_TO_POLL = 3 * 60
+
+export const EMPTY_FUNNEL_RESULTS = {
+    results: [],
+    timeConversionResults: {
+        bins: [],
+        average_conversion_time: 0,
+    },
+}
+
+export async function pollFunnel<T = FunnelStep[]>(apiParams: FunnelRequestParams): Promise<FunnelResult<T>> {
+    // Tricky: This API endpoint has wildly different return types depending on parameters.
```

Comment:
Anything we can do to clear this up? cc @neilkakkar

Reply:
Is this because of the `postgres` based loading?

I think we did away with this for Clickhouse. Optimising this would fall under https://github.com/PostHog/posthog/issues/5313 - doing away with polling.

Reply:
+1 for doing away with polling 🤞

Reply:
Going through these comments one by one—I'll resolve this comment to be addressed under #5313